ambry-router/src/main/java/com.github.ambry.router/CryptoJobHandler.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.utils.Utils;
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Responsible for handling crypto jobs. {@link Router} instantiates this {@link CryptoJobHandler} and gives a
+ * reference of the same to PutManager, GetBlobOperation and GetBlobInfoOperation which will add jobs via
+ * {@link #submitJob(CryptoJob)}. On close, all pending jobs will be processed(either successfully or exception is set)
+ * and any new jobs submitted after close will be ignored.
+ */
+class CryptoJobHandler implements Closeable {
+  private final BlockingQueue<CryptoJob> jobQueue;
+  private final List<CryptoWorker> workerThreads;
+  private AtomicBoolean enabled;
+  private AtomicInteger jobCount = new AtomicInteger(0);
+
+  private static final Logger logger = LoggerFactory.getLogger(CryptoJobHandler.class);
+
+  /**
+   * Instantiates {@link CryptoJobHandler} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   */
+  CryptoJobHandler(CryptoService cryptoService, KeyManagementService kms, int workerCount) {
+    this.jobQueue = new LinkedBlockingQueue<>();
+    this.workerThreads = new ArrayList<>();
+    for (int i = 0; i < workerCount; i++) {
+      workerThreads.add(new CryptoWorker(jobQueue, cryptoService, kms, "CryptoThread_" + i));
+    }
+    this.enabled = new AtomicBoolean(false);
+  }
+
+  /**
+   * Starts up {@link CryptoJobHandler} by spinning up multiple {@link CryptoWorker} threads to assist in executing
+   * {@link CryptoJob}s submitted to {@link CryptoJobHandler} via {@link #submitJob(CryptoJob)}
+   */
+  void startup() {
+    enabled.set(true);
+    for (CryptoWorker worker : workerThreads) {
+      Utils.newThread(worker, true).start();
+    }
+  }
+
+  /**
+   * Submits new job to the {@link CryptoJobHandler}
+   * @param cryptoJob the {@link CryptoJob} that needs to be executed
+   */
+  void submitJob(CryptoJob cryptoJob) {
+    if (enabled.get()) {
+      jobQueue.add(cryptoJob);
+    }
```

Comment:
```
else { 
  cryptoJob.closeJob(closed exception)
}
```

Reply:
Nope. This was intentional to ignore new jobs. Just to be in line with ExecutorService